### PR TITLE
Fixed ALC bug allowed decimal detector groupings

### LINKED ISSUE #33283: ALC crash with bad group input
In the ALC interface giving a custom grouping with a decimal value causes a crash when pressing load. It should give an error (like letters do)

### PR DESCRIPTION
**Pull request by AnthonyLim23**
**Description of work.**

The ALC was allowing users to enter a custom grouping with a decimal detector ID. This caused Mantid to crash. 

**To test:**

Open ALC
Set the instrument to MUSR
In the load bar put 62260
Press the load button - it should be fine

In the grouping section change from auto to custom
In the forward box put "1,2"
In the backward box put "3,4"
Press the load button - it should be fine

In the forward box put "1,2000" (out of range)
In the backward box put "3,4"
Press the load button - it should give a warning pop up 
The message in the load section will be read and say error

In the forward box put "1,2" (out of range)
In the backward box put "a,4"
Press the load button - it should give a warning pop up 
The message in the load section will be read and say error

In the forward box put "1.5 ,2"
In the backward box put "a,4"
Press the load button - it should give a warning pop up 
The message in the load section will be read and say error

Fixes #33283. 

In release notes

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp [hunk 1]
```diff
@@ -30,6 +30,8 @@ namespace {
 const int RUNS_WARNING_LIMIT = 200;
 // must include the "."
 const std::vector<std::string> ADDITIONAL_EXTENSIONS{".nxs", ".nxs_v2", ".bin"};
+
+bool is_decimal(const char character) { return (character == '.') ? true : false; };
```
Comment (AnthonyLim23) [inline qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp:34]:
```suggestion
bool is_decimal(const char character) { return (character == '.') ? true : false; }
```

File: qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp [hunk 2]
```diff
@@ -30,6 +30,8 @@ namespace {
 const int RUNS_WARNING_LIMIT = 200;
 // must include the "."
 const std::vector<std::string> ADDITIONAL_EXTENSIONS{".nxs", ".nxs_v2", ".bin"};
+
+bool is_decimal(const char character) { return (character == '.') ? true : false; }
```
Comment (gui-co) [inline qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp:34]: `return character == '.'` is sufficient.

**Review by gui-co — COMMENTED** [review]
The examples you provided are working well. But here you are only checking that the grouping string starts with a digit and contains no letters and no '.'. It means that anything like "1;2", "1*"... causes Matid to crash. 
I do not know all the use cases, but would it not be better to check if the string contains only digits and commas?

**Comment by AnthonyLim23** [comment]
I only found this bug, because I accidentally pressed "."  with it being next to ","

I don't think users are likely to try any other special characters

File: qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp (hunk 2 above)
Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp:34]:
```suggestion
bool is_decimal(const char character) { return character == '.'; }
```

**Comment by gui-co** [comment]
The real bug seems to be: "entering a wrong character in the field causes Mantid to crash" and this PR does not fix it. In my opinion it would be better to take this opportunity to properly check the value; especially when you risk a hard crash of the whole application.

You know better than me the use cases of that field. If these limitations are fine with you, I approve the PR and let the gatekeeper decide.

**Comment by AnthonyLim23** [comment]
It should be fine. This was to address the accidental pressing of "." when people are tying quickly and miss ","

**Review by gui-co — APPROVED** [review]
As discussed above, I approve the PR because it fixes the bug described in the issue. Adding a decimal point in the lineedit fields no longer crashes Mantid. Tests coverage has also been increased.

In my opinion, this is only fixing a tiny part of the real bug which is that the contents of the fields are not properly checked. Any non-alphabetic and non-numeric character entered in this fields will still cause the same crash of Mantid.

I let the gatekeeper decide on the fate of this PR.

File: qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp [hunk 3]
```diff
@@ -419,7 +421,8 @@ bool ALCDataLoadingPresenter::checkCustomGrouping() {
  * @returns :: True if grouping OK, false if bad
  */
 std::string ALCDataLoadingPresenter::isCustomGroupingValid(const std::string &group, bool &isValid) {
-  if (!std::isdigit(group[0]) || std::any_of(std::begin(group), std::end(group), ::isalpha)) {
```
Comment (martyngigg) [inline qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp:422]: Forgive me as I'm likely not seeing it but would
```cpp
if(!std::any_of(std::begin(group), std::end(group), ::isdigit) {
  isValid = false;
  return "";
}
```
not be what is required?

Comment (AnthonyLim23) [reply qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp:422]: it can have commas in the list

Comment (martyngigg) [reply qt/scientific_interfaces/Muon/ALCDataLoadingPresenter.cpp:422]: Thanks.

**Comment by martyngigg** [comment]
I'll merge this as a focused, low-risk fix for the release.

**Review by DavidFair — APPROVED** [review]

**Comment by DavidFair** [comment]
Ah Martyn, beat me to gatekeeper-ing this :smile:

**Pull request merged**